operator/internal/controller/pulumi/program_controller.go

```diff
@@ -0,0 +1,314 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pulumi
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/opencontainers/go-digest"
+	pulumiv1 "github.com/pulumi/pulumi-kubernetes-operator/operator/api/pulumi/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	ProgramControllerName = "program-controller"
+	pulumiProjectFileName = "Pulumi.yaml"
+)
+
+// ProgramReconciler reconciles a Program object
+type ProgramReconciler struct {
+	client.Client
+	Scheme         *runtime.Scheme
+	Recorder       record.EventRecorder
+	ProgramHandler *ProgramHandler
+}
+
+// ProjectFile contains additional 'project' metadata fields required for Pulumi to run on a ProgramSpec.
+type ProjectFile struct {
+	Name    string `json:"name"`
+	Runtime string `json:"runtime"`
+	pulumiv1.ProgramSpec
+
+	timeModified metav1.Time // this field is not to be serialized
+}
+
+// tarProgram creates a tarball of the Pulumi project file and writes it to the provided writer.
+func tarProgram(data []byte) ([]byte, error) {
```

Comment:
Nit: you should be able to chain the gzip writer instead of using a second buffer. I don't _think_ the info header needs to know the size upfront. Something like this

```go
gzw := gzip.NewWriter(buf)
defer gzw.Close()
tw := tar.NewWriter(gzw)
defer tw.Close()
```

But these files are so small it doesn't really matter.

Edit: and if you keep it all in terms of readers/writers you can use the `ResponseWriter` directly so no need for a buffer.

Reply:
Updated to just combine the 2 functions by chaining the gzip/tar writers and use read/writers instead of a buffer.